src/ahp/constraint_systems.rs

```diff
@@ -91,165 +91,70 @@ pub(crate) fn pad_input<F: PrimeField, CS: ConstraintSystem<F>>(
-    }
+    let max = *[a_density, b_density, c_density]
+        .iter()
+        .max()
+        .expect("iterator is not empty");
+    max
 }
 
-impl<ConstraintF: Field> ConstraintSystem<ConstraintF> for IndexerConstraintSystem<ConstraintF> {
-    type Root = Self;
-
-    #[inline]
-    fn alloc<F, A, AR>(&mut self, _: A, _: F) -> Result<Variable, SynthesisError>
-    where
-        F: FnOnce() -> Result<ConstraintF, SynthesisError>,
-        A: FnOnce() -> AR,
-        AR: Into<String>,
-    {
-        // There is no assignment, so we don't invoke the
-        // function for obtaining one.
-
-        let index = self.num_witness_variables;
-        self.num_witness_variables += 1;
-
-        Ok(Variable::new_unchecked(VarIndex::Aux(index)))
-    }
-
-    #[inline]
-    fn alloc_input<F, A, AR>(&mut self, _: A, _: F) -> Result<Variable, SynthesisError>
-    where
-        F: FnOnce() -> Result<ConstraintF, SynthesisError>,
-        A: FnOnce() -> AR,
-        AR: Into<String>,
-    {
-        // There is no assignment, so we don't invoke the
-        // function for obtaining one.
-
-        let index = self.num_input_variables;
-        self.num_input_variables += 1;
-
-        Ok(Variable::new_unchecked(VarIndex::Input(index)))
-    }
-
-    fn enforce<A, AR, LA, LB, LC>(&mut self, _: A, a: LA, b: LB, c: LC)
-    where
-        A: FnOnce() -> AR,
-        AR: Into<String>,
-        LA: FnOnce(LinearCombination<ConstraintF>) -> LinearCombination<ConstraintF>,
-        LB: FnOnce(LinearCombination<ConstraintF>) -> LinearCombination<ConstraintF>,
-        LC: FnOnce(LinearCombination<ConstraintF>) -> LinearCombination<ConstraintF>,
-    {
-        self.a.push(Self::make_row(&a(LinearCombination::zero())));
-        self.b.push(Self::make_row(&b(LinearCombination::zero())));
-        self.c.push(Self::make_row(&c(LinearCombination::zero())));
-
-        self.num_constraints += 1;
-    }
-
-    fn push_namespace<NR, N>(&mut self, _: N)
-    where
-        NR: Into<String>,
-        N: FnOnce() -> NR,
-    {
-        // Do nothing; we don't care about namespaces in this context.
-    }
+pub(crate) fn make_matrices_square_for_indexer<F: Field>(cs: &mut ConstraintSystem<F>) {
+    let num_variables = cs.num_inputs + cs.num_aux;
+    let nnz = num_non_zero(cs);
+    let matrix_dim = padded_matrix_dim(num_variables, cs.num_constraints);
+    make_matrices_square(cs, num_variables);
+    assert_eq!(
+        cs.num_inputs + cs.num_aux,
+        cs.num_constraints,
+        "padding failed!"
+    );
+    assert_eq!(
+        cs.num_inputs + cs.num_aux,
+        matrix_dim,
+        "padding does not result in expected matrix size!"
+    );
+    assert_eq!(
+        num_non_zero(cs),
+        nnz,
+        "padding changed matrix density"
+    );
+}
 
-    fn pop_namespace(&mut self) {
-        // Do nothing; we don't care about namespaces in this context.
-    }
+pub(crate) fn make_matrices_square_for_prover<F: Field>(cs: &mut ConstraintSystem<F>) {
```

Comment:
Do we really need 2 separate functions for indexer and prover ? It seems to me that they do exactly the same thing, but the indexer function has just a couple of additional checks